File: Graphs/floyd_warshall_shortest_path.cpp
/**
 * @Author: hardcodecoder
 * @Date:   04:24:42 Thursday 16 April 2020
 * @Last modified by:   hardcodecoder
 * @Last modified time: 06:12:12 Thursday 16 April 2020
 */


#include<stdio.h>
#include<stdlib.h>

#define INF 9999


// Stores the number of vertices in the graph
int numVertices;

// Adjacency matrix to store the
// shortest distances from each vertex
int **shortestDistanceMatrix;


// Method to print matrix
void printMatrix() {
  int i,j;
  printf ("\nThe following matrix shows the shortest distances between every pair of vertices \n");
  for (i = 0; i < numVertices; i++) {
    for (j = 0; j < numVertices; j++) {
      if (shortestDistanceMatrix[i][j] == INF)
        printf("%5s", "INF");
      else
        printf("%5d", shortestDistanceMatrix[i][j]);
    }
    printf("\n");
  }
}


// Implementation of Floyd-Marshall shortest path algorithm
void floydWarshallShortestPath(int **matrix) {
  int i, j, k;

  // Allocalte memory for pointer to pointer for 2D array
  shortestDistanceMatrix = (int**) malloc(sizeof(int*) * numVertices);

  // Initialize the output matrix same as the source matrix
  for (i = 0; i < numVertices; i++){
    // Allocate memory for each row
    shortestDistanceMatrix[i] = (int*) malloc(sizeof(int) * numVertices);
    for (j = 0; j < numVertices; j++)
        shortestDistanceMatrix[i][j] = matrix[i][j];
  }

  // Iterate with kth vertex as the intermediate vertex
  for (k = 0; k < numVertices; k++){

    // Select all vertices as source for each kth intermediate vertex
    for (i = 0; i < numVertices; i++){

      // Select each vertex as destination for each ith vertex
      for (j = 0; j < numVertices; j++){
        if (shortestDistanceMatrix[i][k] != INF && shortestDistanceMatrix[k][j] != INF &&
           shortestDistanceMatrix[i][j] > shortestDistanceMatrix[i][k] + shortestDistanceMatrix[k][j])
          shortestDistanceMatrix[i][j] = shortestDistanceMatrix[i][k] + shortestDistanceMatrix[k][j];
      }
    }
  }

  // If distance of any verex from itself becomes
  // negative, then there is a negative weight cycle.
  for (i = 0; i < numVertices; i++){
    if (shortestDistanceMatrix[i][i] < 0){
      printf("Negative edge detected, terminating...");
      return;
    }
  }
  // Print the final output
  printMatrix();
}



int main(){
  int **graph;

  printf("Enter the number of vertex : ");
  scanf("%d", &numVertices);
  printf("\nEnter the adjacency matrix (9999 for infinity):\n");

  // pointer to pointer for storing 2d array
  graph = (int**)malloc(sizeof(int*)*numVertices);

  for (int i = 0; i < numVertices; i++){
    // allocate memory from each row
    graph[i] = (int*)malloc(sizeof(int)*numVertices);

    // fill each coloum of ith row
    for (int j = 0; j < numVertices; j++)
      scanf("%d", &graph[i][j]);//or use (*(graph+i)+j)
  }

  floydWarshallShortestPath(graph);
  /* Test input
  0 5 9999 10
  9999 0 3 9999
  9999 9999 0 1
  9999 9999 9999 0
  */
  return 0;
}

File: Graphs/Java/FloydWarshallShortestPath.java
/**
 * @Author: hardcodecoder
 * @Date:   03:10:18 Friday 15 May 2020
 * @Last modified by:   hardcodecoder
 * @Last modified time: 03:42:45 Friday 15 May 2020
 */


 import java.io.*;
 import java.util.Scanner;

 public class FloydWarshallShortestPath {

   // Defines  distance for unreachable vertex
   private static final float INF = 9999;

   // Stores the final shortest distances from source to every other vertex
   // Index represent the vertex no. from source and the value gives the distance
   private float[][] shortestDistance;

   // Main method
   public static void main(String args[]) {
     new FloydWarshallShortestPath().initialize();
   }

   // Helper method to input data and initialize necessary variables
   private void initialize() {
     int numVertex;

     System.out.println("Enter the number of vertices: ");
     Scanner sc = new Scanner(System.in);
     numVertex = sc.nextInt();

     // Initialize all necessary variables
     float[][] adjacencyMatrix = new float[numVertex][numVertex];
     shortestDistance = new float[numVertex][numVertex];

     // Input the adjacency matrix also,
     // Initialize the output matrix as input matrix
     System.out.println("Enter the adjacency matrix: ");
     for (int i = 0; i < numVertex; i++) {
       for (int j = 0; j < numVertex; j++) {
         float f = sc.nextFloat();
         adjacencyMatrix[i][j] = f;
         shortestDistance[i][j] = f;
       }
     }

     findShortestpath(numVertex);
   }

   // Implementation of floyd-warshall algorithm
   private void findShortestpath(int numVertex) {
     // Iterate with kth vertex as intermediate vertex
     for (int k = 0; k < numVertex; k++) {
       // Iterate with ith vertex as source for kth intermediate vertex
       for (int i = 0; i < numVertex; i++) {
         // Iterate with jth vertex as destination vertex for
         // ith source vertex and kth intermediate vertex
         for (int j = 0; j < numVertex; j++) {
           // Find shortest path from i-->j with k as intermediate vertex
           if (shortestDistance[i][k] != INF && shortestDistance[k][j] != INF
            &&shortestDistance[i][j] > shortestDistance[i][k] + shortestDistance[k][j])
                  shortestDistance[i][j] = shortestDistance[i][k] + shortestDistance[k][j];
         }
       }
     }

     // If distance of any verex from itself becomes
     // negative, then there is a negative weight cycle.
     for (int i = 0; i < numVertex; i++){
       if (shortestDistance[i][i] < 0){
         System.out.println("Negative edge detected, terminating...");
         return;
       }
     }

     print(numVertex);
   }

   // Helper method to print the output matrix
   private void print(int numVertex) {
     System.out.println("The following matrix shows the shortest distances between every pair of vertices:");
     for (int i = 0; i < numVertex; i++) {
       for (int j = 0; j < numVertex; j++) {
         if (shortestDistance[i][j] == INF)
           System.out.print(String.format("%5s", "INF"));
         else
           System.out.print(String.format("%5.1f", shortestDistance[i][j]));
       }
       System.out.println();
     }
   }
 }
